From a204c2e45ed26f5ae270998a669337e2262f2e54 Mon Sep 17 00:00:00 2001
From: Chuanxiao Dong <chuanxiao.dong@intel.com>
Date: Fri, 27 Feb 2015 16:21:49 +0800
Subject: [PATCH] mmc: clear data if error happens

With some buggy storage device, after receiving CMD error interrupt
storage device may continue to send DATA error interrupt, which can
make driver to handle this DATA error interrupt. In this case, driver
can clear the data if CMD error happens.

Before start a new request in error case, also needs to clear previous
data to release the resource

Change-Id: I906e4f14afe918c1b5092b3ea7e36f6b7cc93cc6
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-23214
Signed-off-by: Chuanxiao Dong <chuanxiao.dong@intel.com>
---
 drivers/mmc/card/block.c | 3 +--
 drivers/mmc/host/sdhci.c | 2 ++
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/card/block.c b/drivers/mmc/card/block.c
index b607175..3497970 100644
--- a/drivers/mmc/card/block.c
+++ b/drivers/mmc/card/block.c
@@ -2039,10 +2039,9 @@ static int mmc_blk_issue_normal_rw_rq(struct mmc_queue *mq, struct request *rqc)
 			ret = blk_end_request(req, -EIO,
 					blk_rq_cur_bytes(req));
 	}
+ start_new_req:
 	clear_bit_unlock(mq_rq->task_id, &mq->cmdqslot);
 	atomic_dec(&mq->active_slots);
-
- start_new_req:
 	if (rqc) {
 		if (mmc_card_removed(card)) {
 			rqc->cmd_flags |= REQ_QUIET;
diff --git a/drivers/mmc/host/sdhci.c b/drivers/mmc/host/sdhci.c
index faa1706..69003bd 100644
--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -2354,6 +2354,8 @@ static void sdhci_tasklet_finish(unsigned long param)
 		   controllers do not like that. */
 		sdhci_reset(host, SDHCI_RESET_CMD);
 		sdhci_reset(host, SDHCI_RESET_DATA);
+		/* clear data as DATA is reset */
+		host->data = NULL;
 	}
 
 	host->mrq = NULL;
-- 
1.9.1

